src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Catalog/PackageHasSignatureValidator.cs

```diff
@@ -0,0 +1,140 @@
+                    data.Package.Id,
+                    data.Package.Version);
+
+                return Task.FromResult(false);
+            }
+
+            return Task.FromResult(true);
+        }
+
+        public async Task RunValidatorAsync(ValidationContext data)
+        {
+            // Ensure each package on the catalog has a signature file.
+            foreach (var entry in data.Entries)
+            {
+                // We don't need to validate deletion entries.
+                if (entry.IsDelete())
+                {
+                    _logger.LogInformation(
+                        "Skipping deletion catalog entry {CatalogEntry} for package {PackageId} {PackageVersion}",
+                        entry.Uri,
+                        data.Package.Id,
+                        data.Package.Version);
+
+                    continue;
+                }
+
+                _logger.LogInformation(
+                    "Validating that catalog entry {CatalogEntry} for package {PackageId} {PackageVersion} has a package signature file...",
+                    entry.Uri,
+                    data.Package.Id,
+                    data.Package.Version);
+
+                var leaf = await data.Client.GetJObjectAsync(entry.Uri, data.CancellationToken);
+
+                if (!LeafHasSignatureFile(leaf))
+                {
+                    _logger.LogWarning(
+                        "Catalog entry {CatalogEntry} for package {PackageId} {PackageVersion} is missing a package signature file",
+                        entry.Uri,
+                        data.Package.Id,
+                        data.Package.Version);
+
+                    throw new MissingPackageSignatureFileException(
+                        entry.Uri,
+                        data.Package.Id,
+                        data.Package.Version.ToNormalizedString(),
+                        $"Package {data.Package.Id} {data.Package.Version} is missing a package signature file");
+                }
+
+                _logger.LogInformation(
+                    "Validated that catalog entry {CatalogEntry} for package {PackageId} {PackageVersion} has a package signature",
+                    entry.Uri,
+                    data.Package.Id,
+                    data.Package.Version);
+            }
+        }
+
+        private bool LeafHasSignatureFile(JObject leaf)
+        {
+            var packageEntries = leaf["packageEntries"];
+
+            if (packageEntries == null)
+            {
+                return false;
+            }
+
+            if (!(packageEntries is JArray files))
+            {
+                return false;
+            }
+
+            foreach (var file in files)
+            {
+                var path = (string)file["fullName"];
+
+                if (path == NupkgSignatureFile)
+                {
+                    return true;
+                }
+            }
+
+            return false;
```

Comment:
You could simplify lines 127-137 to:

    return files.Any(file => file.ToString() == NupkgSignatureFile);